File: cpp/algorithms/advanced/sqrt_decomposition.cpp
#include <bits/stdc++.h>
using namespace std;

struct SqrtDecomposition {
    vector<int> a;
    vector<int> blk;
    int n, b;
    explicit SqrtDecomposition(vector<int> arr) : a(move(arr)) {
        n = (int)a.size();
        b = max(1, (int)sqrt(n));
        for (int i = 0; i < n; i += b) {
            int s = 0;
            for (int j = i; j < min(n, i + b); ++j) s += a[j];
            blk.push_back(s);
        }
    }
    int id(int i) const { return i / b; }
    void point_add(int i, int delta) {
        a[i] += delta;
        blk[id(i)] += delta;
    }
    int range_sum(int l, int r) const {
        int s = 0;
        if (id(l) == id(r)) {
            for (int i = l; i <= r; ++i) s += a[i];
            return s;
        }
        while (l <= r && l % b != 0) s += a[l++];
        while (l <= r && (r + 1) % b != 0) s += a[r--];
        while (l <= r) {
            s += blk[id(l)];
            l += b;
        }
        return s;
    }
};

int main() {
    SqrtDecomposition sd({1, 2, 3, 4, 5});
    assert(sd.range_sum(0, 4) == 15);
    sd.point_add(2, 10);
    assert(sd.range_sum(0, 4) == 25);
    assert(sd.range_sum(1, 3) == 2 + 13 + 4);
    cout << "advanced OK" << endl;
    return 0;
}

File: python/algorithms/advanced/sqrt_decomposition.py
"""分块（Sqrt decomposition）：区间和查询 + 单点加。"""

from __future__ import annotations
import math


class SqrtDecomposition:
    """将数组分块，块内维护元素和；单点加 O(1)，区间和 O(√n)。"""

    def __init__(self, a: list[int]) -> None:
        self.a = a[:]
        self.n = len(self.a)
        self.b = max(1, int(math.sqrt(self.n)))
        self._build_blocks()

    def _build_blocks(self) -> None:
        self.blk: list[int] = []
        for i in range(0, self.n, self.b):
            self.blk.append(sum(self.a[i : i + self.b]))

    def _id(self, i: int) -> int:
        return i // self.b

    def point_add(self, i: int, delta: int) -> None:
        self.a[i] += delta
        self.blk[self._id(i)] += delta

    def range_sum(self, l: int, r: int) -> int:
        s = 0
        if self._id(l) == self._id(r):
            return sum(self.a[l : r + 1])
        while l <= r and l % self.b != 0:
            s += self.a[l]
            l += 1
        while l <= r and (r + 1) % self.b != 0:
            s += self.a[r]
            r -= 1
        while l <= r:
            s += self.blk[self._id(l)]
            l += self.b
        return s


if __name__ == "__main__":
    sd = SqrtDecomposition([1, 2, 3, 4, 5])
    assert sd.range_sum(0, 4) == 15
    sd.point_add(2, 10)
    assert sd.range_sum(0, 4) == 25
    assert sd.range_sum(1, 3) == 2 + 13 + 4
    print("advanced OK")
